### PR TITLE
Added Fluid object references to the result of summarize to be used in garbage collection

### LINKED ISSUE #4402: Update summarize to also return a list of references to fluid objects.
Summarize should also return a list of references to fluid objects. This will be used to form the "reference graph" that garbage collection will run on.

### PR DESCRIPTION
**Pull request by agarwal-navin**
Fixes #4402.

Added a list of garbage collection nodes to be returned along with the summarize result. Each fluid object that can summarize (data stores, dds for now) is considered a "node" and will return this list.
The definition of a garbage collection node:
```
interface IGraphNode
{
    id: string;
    outboundRoutes: string[];
}
```
`id` is an identifier for this node. Each node returns "/" for its own id. The parent adds the child's id to the `id`.
`outboundRoutes` is a set of routes to other garbage collection nodes.

**Key things to note:**
- Each node returns the GC nodes of its children nodes plus its own GC node.
  - Currently, a data store returns all of its channel contexts (DDS) as outbound routes. This will change to returning only "root" DDS once we have root and non-root DDS.
  - ContainerRuntime will returns all root data stores as its outbound routes (this will be done is a subsequent PR that runs GC algorithm).
- Added a derived class of SummarizerNode that caches this list of references at each node. If nothing changed since last summary, it uses the cached list.
  - It also adds "trackState" to its summarize method so that the caller doesn't have to handle that scenario.
  - Added a method that the SummarizerNode can call to get the initial GC nodes as per the last summary.
  - Todo: The part of adding the GC nodes to summary and fetching it back will be done in a subsequent PR.

### CODE REVIEW

File: packages/runtime/datastore/src/localChannelContext.ts [hunk 1]
```diff
@@ -124,23 +123,26 @@ export class LocalChannelContext implements IChannelContext {
...
      */
-    public async summarize(fullTree: boolean = false, trackState: boolean = false): Promise<ISummarizeResult> {
+    public async summarize(fullTree: boolean = false, trackState: boolean = false): Promise<IContextSummarizeResult> {
         assert(this.isLoaded && this.channel !== undefined, "Channel should be loaded to take summary");
         const snapshot = snapshotChannel(this.channel);
         const summary = convertToSummaryTree(snapshot, fullTree);
-        return summary;
+        return {
+            ...summary,
+            references: [],
```
Comment (agarwal-navin) [inline packages/runtime/datastore/src/localChannelContext.ts:132]: This is temporary until #3975 is merged. Then the actual references returned by DDS will be used.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 2]
```diff
@@ -563,12 +567,36 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
         return tree.entries;
     }
 
+    /**
+     * Returns the Fluid objects referenced by this channel.
+     */
+    private getFluidObjectReferences(): IFluidObjectReferences {
+        /**
+         * Get the Fluid objects referenced by this channel. Currently, all contexts in this channel are considered
+         * as referenced. This will change will we have root and non-root channel contexts. Then only root contexts
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:576]: "This will change will we have"  - can't parse :)

File: packages/runtime/runtime-definitions/src/summary.ts [hunk 3]
```diff
@@ -11,6 +11,21 @@ import {
     ITree,
 } from "@fluidframework/protocol-definitions";
 
+/**
+ * Represents routes from a Fluid object to other Fluid objects in the Container. This is used for garbage collection.
+ */
+export interface IFluidObjectReferences {
```
Comment (vladsud) [inline packages/runtime/runtime-definitions/src/summary.ts:17]: I'd call it a "GC node". its path property is a route, and routes represents outbound connections.
But that thing combined is node, it's not "references". It represents single node for GC system

File: packages/runtime/runtime-utils/src/fluidObjectReferencesUtils.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    return clonedReferences;
+}
+
+/**
+ * Helper function that adds a route to the routes of given Fluid object references.
+ * @param references - The list of Fluid object references to add the route to.
+ * @param route - The route to be added to the Fluid object references.
+ */
+export function addRouteToReferences(references: IFluidObjectReferences[], route: string) {
```
Comment (vladsud) [inline packages/runtime/runtime-utils/src/fluidObjectReferencesUtils.ts:47]: addRouteTo AllNodes?

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 5]
```diff
@@ -563,12 +567,36 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
         return tree.entries;
     }
 
+    /**
+     * Returns the Fluid objects referenced by this channel.
+     */
+    private getFluidObjectReferences(): IFluidObjectReferences {
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:573]: Based on what this API does and how it's used, I'd call it getGcNodeForDataStore(), or simply getGcNode() (given that's API on data store). It returns single node (with properties describing this node), not references.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 6]
```diff
@@ -581,11 +609,31 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
...
+                // back-compat 0.30 - Older versions will not return references. Set it to empty array.
+                if (contextSummary.references === undefined) {
+                    contextSummary.references = [];
+                }
+                // Add the context's Fluid object references to the list of this channel's references.
+                references.push(...contextSummary.references);
             }));
 
-        return builder.getSummaryTree();
+        // Normalize all the context's path of Fluid object references and prefix them with this channel's id.
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:623]: Some comment here would be useful to describe higher-level intention, as it took me a while to get it.
The intention is that we create (for now) circular dependencies between data store and channels, such that they are referenced or collected all at once. In future, we may change that to allow individual DDS (or even maybe sub-DDS) collection.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 7]
```diff
@@ -610,22 +661,41 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
...
                 }
-
-                builder.addWithStats(objectId, summary);
+                builder.addWithStats(objectId, summaryTree);
             }
         }
 
-        return builder.getSummaryTree();
+        // Normalize all the context's path of Fluid object references and prefix them with this channel's id.
+        normalizeAndPrefixReferencesPath(references, this.id);
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:686]: can we please reduce code duplication? And reuse code between these two summarization methods?

File: packages/runtime/runtime-definitions/src/summary.ts [hunk 8]
```diff
@@ -23,16 +38,25 @@ export interface ISummaryTreeWithStats {
...
+export interface IChannelSummarizeResult extends ISummaryTreeWithStats {
+    references: IFluidObjectReferences[];
+}
+
 export interface ISummarizeResult {
     stats: ISummaryStats;
     summary: SummaryTree;
 }
 
-export interface ISummarizeInternalResult extends ISummarizeResult {
```
Comment (vladsud) [inline packages/runtime/runtime-definitions/src/summary.ts:31]: ISummaryTreeWithStats  and ISummarizeResult  are the same
hence IChannelSummarizeResult  & IContextSummarizeResult  are the same.
Can we merge them?

File: packages/runtime/runtime-utils/src/fluidObjectReferencesUtils.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        reference.path = `${prefixId}/${normalizedPath}`;
+    }
+}
+
+/**
+ * Helper function that clones the given list of Fluid object references.
+ * @param references - The list of Fluid object references to be cloned.
+ * @returns - A clone of the given Fluid object references.
+ */
+export function cloneFluidObjectReferences(references: IFluidObjectReferences[]): IFluidObjectReferences[] {
```
Comment (vladsud) [inline packages/runtime/runtime-utils/src/fluidObjectReferencesUtils.ts:31]: does not look specialization is needed here. Can we use deepClone()?
I do not remember where we stand re using lodash - I think Curtis does not like it, but we use it :)

File: packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts [hunk 10]
```diff
@@ -0,0 +1,242 @@
+/*!
```
Comment (vladsud) [inline packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts:1]: That's a lot of new code, and no UTs? :)

File: packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+    ): ISummarizerNodeWithReferences {
+        assert(!this.children.has(id), "Create SummarizerNode child already exists");
+
+        const latestSummary = this.latestSummary;
+        let child: SummarizerNodeWithReferences;
+        switch (createParam.type) {
+            case CreateSummarizerNodeSource.FromAttach: {
+                let summaryNode: SummaryNode | undefined;
+                let initialSummary: IInitialSummary | undefined;
+                if (
```
Comment (vladsud) [inline packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts:154]: I only briefly looked through this code, so maybe my question is not valid, but still:
Any chance we can not have in this class logic around sequence numbers? I.e. can base class do all the heavy lifting and provide right accessors, and this object only be responsive for reading / writing GC info?
I worry about always increasing complexity of summarizing domain, and logic being spread across so many layers, that when things go wrong, we are left completely clueless on how to approach it.

**Comment by vladsud** [comment]
Overall looks good to me, but naming & UT domain needs improvements :)

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:576]: oops :-).

File: packages/runtime/runtime-definitions/src/summary.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-definitions/src/summary.ts:17]: I went back and forth with this (naming is hard!). The reason I didn't add garbage collection or GC phrase to this name is because I wanted to keep it independent of GC. The data stores and channel contexts simply return a set of its references. Then a root entity will use this and pass it to the GC process as an IGarbageCollectionNode. How about IReferenceGraphNode - It is used to build a reference graph of fluid objects in this document?

Having said that, I don't have very strong feeling towards it and calling it IGarbageCollectionNode is fine with me :-).

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:623]: Actually, that's not what this is doing. This is just adding its "id" as a prefix to the path of child's node.
For example, let's say this data store's id is "default" and its child is a DDS "foo".
The nodes returned by "foo" will look something like this -
{ path: "foo", routes: ["/dataStore2"] }
This line will update the path by adding its id as prefix, so this becomes:
{ path: "default/foo", routes: ["/dataStore2"] }

Basically, each layer adds its id as prefix to the path of its children nodes before returning. This will make sure that that "path" we get at the root is the full path.

Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:623]: The "addRouteToReferences" call below adds a back reference to itself for all the children so that if a child is referenced, so is this parent.

File: packages/runtime/runtime-definitions/src/summary.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-definitions/src/summary.ts:31]: ISummaryTreeWithStats and ISummarizeResult differ in that ISummarizeResult's summary can also be a ISummaryHandle. ISummaryTreeWithStats's summary is always an ISummaryTree.
We can probably remove ISummaryTreeWithStats and assert that channel's always return an ISummaryTree and never an ISummaryHandle.

File: packages/runtime/runtime-utils/src/fluidObjectReferencesUtils.ts (hunk 9 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/fluidObjectReferencesUtils.ts:31]: I did use deepClone() at some point in time. But it gave me an error and I had to disable an eslint warning. So, that got me thinking that we shouldn't be using it unless necessary. The clone needed here seemed straightforward so I just implemented it.

File: packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts (hunk 10 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts:1]: The only new code is adding the fluid object references. Rest if duplicate of SummarizerNode. I updated dataStoreContext.spec.ts and dataStoreCreation.spec.ts to use this new class and it worked fine.
I will add some tests that verify we get correct references.

File: packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts (hunk 11 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts:154]: I totally agree. The thing is SummarizerNode is not designed to be inherited :-). So, I will have to make changes to it such that it can be inherited without requiring leaking the core logic. I will look into it.

File: packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts (hunk 10 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts:1]: Added UTs for the added functionality.

File: packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts (hunk 11 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/summarizerNodeWithReferences.ts]: Restructured the code a little to do this. Can you please take a look and let me know.

**Review by vladsud — APPROVED** [review]
I briefly look at last commits - would love to see more UTs for the layer you've added, otherwise looks good.

File: packages/runtime/runtime-definitions/src/summary.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-definitions/src/summary.ts]: Changed it to IGraphNode. The variables / members are called gcNodes.

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: Updated the code and added comments to make this clearer.

File: packages/runtime/runtime-utils/src/fluidObjectReferencesUtils.ts (hunk 9 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/fluidObjectReferencesUtils.ts]: I spoke to Curtis and it seems like loadash increases the package size a lot. We may plan to remove it at some point. I will keep this implementation so that I don't add more dependencies.

File: packages/runtime/runtime-definitions/src/summary.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-definitions/src/summary.ts:31]: I will speak to @arinwt about this and consolidate this if possible in a separate PR.

**Comment by agarwal-navin** [comment]
> I briefly look at last commits - would love to see more UTs for the layer you've added, otherwise looks good.

I will keep adding more UTs. Did you have specific scenarios in mind?

File: packages/runtime/runtime-definitions/src/summary.ts (hunk 8 above)
Comment (arinwt) [reply packages/runtime/runtime-definitions/src/summary.ts:31]: I like differing types, but think we could name them better

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 12]
```diff
@@ -1470,36 +1472,87 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         return this.context.submitSignalFn(envelope);
     }
 
+    /**
+     * Updates the garbage collection nodes of this node's children:
+     * - Prefixs the child's id to the id of each node returned by the child.
+     * @param chilGCNodes - The child's garabage collection nodes.
```
Comment (arinwt) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1478]:
```suggestion
     * @param childGCNodes - The child's garabage collection nodes.
```

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 13]
```diff
@@ -1470,36 +1472,87 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
+     * @param childId - The id of the child node.
+     * @returns the updated GC nodes of the child.
+     */
+    private updateChildGCNodes(chilGCNodes: IGraphNode[], childId: string): IGraphNode[] {
```
Comment (arinwt) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1482]:
```suggestion
    private updateChildGCNodes(childGCNodes: IGraphNode[], childId: string): IGraphNode[] {
```

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 14]
```diff
@@ -1470,36 +1472,87 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
+        // Normalize the child's nodes and prefix the child's id to the ids of GC nodes returned by it.
+        // This gradually builds the id of each node to be a path from the root.
+        normalizeAndPrefixGCNodeIds(chilGCNodes, childId);
```
Comment (arinwt) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1485]:
```suggestion
        normalizeAndPrefixGCNodeIds(childGCNodes, childId);
```

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 15]
```diff
@@ -1470,36 +1472,87 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
+        return chilGCNodes;
```
Comment (arinwt) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1486]:
```suggestion
        return childGCNodes;
```

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 13 above)
Comment (arinwt) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1482]: I'm not sure what value does this function add on top of `normalizeAndPrefixGCNodeIds`?  Seems like this could be removed altogether and callers can just call `normalizeAndPrefixGCNodeIds`.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1478]: I need a spell checker! :-)

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 13 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1482]: There are 2 reasons I did it (although not strong reasons) -
1. I might add a reference back to container runtime to all its child nodes. I have not finalized it yet.
2. I want to keep the implementation in dataStoreRuntime and containerRuntime as similar as possible. From GC's perspective, they are the same.

I will remove this method if it doesn't make sense once I have the main GC implementation.

File: packages/runtime/runtime-definitions/src/summary.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-definitions/src/summary.ts:31]: Do you have any names in mind?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 16]
```diff
@@ -1470,36 +1472,87 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
+                const contextSummary = await context.summarize(fullTree, trackState);
+                builder.addWithStats(contextId, contextSummary);
+
+                // back-compat 0.30 - Older versions will not return GC nodes. Set it to empty array.
+                if (contextSummary.gcNodes === undefined) {
+                    contextSummary.gcNodes = [];
+                }
+
+                // Update and add the child context's GC nodes to the main list.
+                gcNodes.push(...this.updateChildGCNodes(contextSummary.gcNodes, contextId));
```
Comment (arinwt) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1542]: Consider using `concat` to avoid the unnecessary spread if we expect number of GC nodes can be large.
Concat will not mutate array, so it can't be const in that case.

File: packages/runtime/runtime-definitions/src/summary.ts (hunk 8 above)
Comment (arinwt) [reply packages/runtime/runtime-definitions/src/summary.ts:31]: No, I think those are fine. I meant the underlying type names are bad at distinguishing themselves: `ISummaryTreeWithStats` and `ISummarizeResult`.

Comment (arinwt) [reply packages/runtime/runtime-definitions/src/summary.ts:31]: I think it boils down to `SummaryTree` being poorly named.
If we had a unique term for what `SummaryTree` currently is: `ISummaryTree | ISummaryHandle`. Then we could use that to distinguish all the way up the stack. For example:
`ISummaryTree` vs. `ISummaryTreeOrHandle`
`ISummaryTreeWithStats` vs. `ISummaryTreeOrHandleWithStats`
etc.
But `ISummaryTreeOrHandle` is pretty long...

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 13 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1482]: I guess that makes sense. I feel like that code doesn't need to live here though. Can you just have a function that calls both `normalizeAndPrefixGCNodeIds` _and_ `addRouteToAllGCNodes` that takes in the absolutePath as a parameter?

Just feel like I wanna keep these classes as slim as possible (though we are already _not_ succeeding with ContainerRuntime).

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 16 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1542]: That's a great suggestion, thanks!

File: packages/runtime/runtime-definitions/src/summary.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-definitions/src/summary.ts:31]: I agree. SummaryTree and ISummaryTree sound very similar and its strange how different they are :-).
Created issue #4434 to track this. Assigned to both you and me. One of us can take this up.

File: packages/runtime/runtime-utils/src/summarizerNode.ts [hunk 17]
```diff
@@ -45,7 +45,7 @@ async function seqFromTree(
 }
 
 /** Path for nodes in a tree with escaped special characters */
-class EscapedPath {
+export class EscapedPath {
```
Comment (arinwt) [inline packages/runtime/runtime-utils/src/summarizerNode.ts:48]: nit: I see you exported these so you could use them in the extended class, but as far as minimizing leaking implementation goes, you may want to somehow filter what gets exported from the package.

So these shared things can be exported from a file: `summarizerNodeCore.ts` (or whatever you choose to name it)
Then the things in this file which were already exported can remain `summarizerNode.ts`
And the things in your new file can be exported `summarizerNodeWithGc.ts`
But the runtime-utils package will not export the core file in `index.ts`.

Does that make sense?

File: packages/runtime/runtime-utils/src/summarizerNodeWithGc.ts [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+            async (fullTree: boolean) => this.summarizeInternal(fullTree, true /* trackState */),
+            config,
+            changeSequenceNumber,
+            latestSummary,
+            initialSummary,
+            wipSummaryLogger,
+        );
+    }
+
+    public async summarize(fullTree: boolean, trackState: boolean = true): Promise<IContextSummarizeResult> {
```
Comment (arinwt) [inline packages/runtime/runtime-utils/src/summarizerNodeWithGc.ts:52]: Curious: why move `trackState` concept into SummarizerNode?

Comment (arinwt) [reply packages/runtime/runtime-utils/src/summarizerNodeWithGc.ts:52]: Nevermind, I get it. It's to always do the GC node work even if `trackState` is false. I misread it the first time.

File: packages/runtime/runtime-utils/src/summarizerNode.ts (hunk 17 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/summarizerNode.ts:48]: Yes, I totally agree with this.
I actually wanted to do something similar. Move out the pieces that are not in the core SummarizerNode class to a separate file, including some helper functions. That would make the summarizerNode.ts file smaller and cleaner.
But I don't want to do this as part of this change since its not related. I would rather do that in a separate PR that might end up cleaning the summarizer node some more.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 13 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: Yes, I agree that we want to keep ContainerRuntime and DataStoreRuntime as slim as possible. But this is not getting added to the IContainerRuntime interface so it shouldn't be a big deal.
Also, I think garbage collection will be core to every layer we have so we will have some GC related functionality.

Maybe the long term solution is to move garbage collection functionality to a separate helper class so the core classes are cleaner.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 19]
```diff
@@ -563,29 +567,83 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
...
+        // Normalize the child's nodes and prefix the child's id to the ids of GC nodes returned by it.
+        // This gradually builds the id of each node to be a path from the root.
+        normalizeAndPrefixGCNodeIds(childGCNodes, childId);
+
+        // Add a back route to self in each child's outbound routes. If any child is referenced, then its parent
+        // should be considered referenced as well.
+        addRouteToAllGCNodes(childGCNodes, this.absolutePath);
+
+        return childGCNodes;
+    }
```
Comment (arinwt) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:588]: Okay so I was thinking about the layering a little bit, and (this is more of a question) I was wondering if `normalizeAndPrefixGCNodeIds` should be pushed _down_ a layer to the individual channel contexts? And similarly push down the one in ContainerRuntime to each DataStoreContext.

I was thinking each node knows its own ID, so it could just return `id/` instead of `/` for itself.

While it feels like there could be something to gain by abstracting each node from its own ID, we already don't do that, and I'm not clearly seeing what benefit we do get in exchange for what looks like logic in the wrong place (to me).

**Comment by arinwt** [comment]
Really appreciate all the renames/refactors along the way and the tests!

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 19 above)
Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:588]: Great observation! I was doing it as you described in one of the previous iterations. I had a discussion with Vlad and changed it.
You are right that each layer today knows its own id and so it can return its own id instead of "/". From what Vlad told me, we will be (mostly) removing the concept of id from each layer since they don't really need to know about it (we do use it in few places but that will change). The id of a layer will be known only to its parent.

This also keeps the design similar to summarize where each layer returns its summary data and the parent stores it against the child's id as key.

Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:588]: I think this is a better design since an object may or may not know its own id but a parent surely knows its children's id. I was also thinking about scenarios where something like a collection component might returns gc nodes for each of its collection instance. There, the instances may not care about its id but the collection does.
So, this design is more generic and should work well in all scenarios.

**Comment by agarwal-navin** [comment]
> Really appreciate all the renames/refactors along the way and the tests!

No problem, happy to clean up as much as I can :-). I would appreciate your help in writing more summarizer node UTs.

File: packages/runtime/runtime-utils/src/summarizerNode.ts (hunk 17 above)
Comment (arinwt) [reply packages/runtime/runtime-utils/src/summarizerNode.ts:48]: Well the only thing is, technically this PR will make that subsequent change a "breaking" change, though we don't need to announce that since (hopefully) nobody will directly reference these internals.

Comment (arinwt) [reply packages/runtime/runtime-utils/src/summarizerNode.ts:48]: But I'm totally fine to do it in separate PR.  Even I can do it if you'd like to make sure we get it in same release.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 13 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: I guess I just mean these `updateChildGCNodes` functions are all roughly the same, which just call the 2 external helper functions. Couldn't that itself be a helper function?

Comment (arinwt) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: Also, not a big deal if it stays here, just feels a little unnecessary and slightly code duplication

File: packages/runtime/runtime-utils/src/summarizerNode.ts (hunk 17 above)
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/summarizerNode.ts:48]: Yes to getting in the same release. Thanks, I would really appreciate if you could help with this :-).

**Review by arinwt — APPROVED** [review]
Good PR, I don't see any problems with the concept/design and would be fine to merge as-is.

Left a few comments on possible different design decisions, i.e. reduce additional logic to existing runtimes/contexts, shift from inheritance to composition for SummarizerNode + GC state tracking/summarization functionality.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 13 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: I guess you are right. If every node is going to do the same thing, it might be worth to add it as a helper. Tell you what, once I have the GC algorithm in place, it will be much more clearer what each node looks like and I can handle it then.
Right now, I am not sure if containerRuntime will need to call both these helpers and if not, I will remove the updateChildGCNodes method.

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 19 above)
Comment (arinwt) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:588]: Thanks for letting me know!  I agree with that direction, and can help to pull it out of SummarizerNode as well.. it was on my mind before as well.

File: packages/runtime/runtime-utils/src/summarizerNode.ts (hunk 17 above)
Comment (arinwt) [reply packages/runtime/runtime-utils/src/summarizerNode.ts:48]: Opened a PR for this #4447

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 13 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: Sounds totally reasonable, I like it! My view is that this is a "big" PR which gets everything working (as needed for subsequent PRs) and then we can do small, incremental PRs with little-to-no functional change once we have a clear big picture.
